Enable multiple calls to mbedtls_gcm_update_ad

mstarzyk-mobica opened the pull request:
Resolves #4351

mstarzyk-mobica commented on tests/suites/test_suite_gcm.function:
```diff
@@ -16,15 +16,19 @@ static int check_multipart( mbedtls_gcm_context *ctx,
     int ok = 0;
     uint8_t *output = NULL;
     size_t n2 = input->len - n1;
+    size_t n1_add = add->len/2;
```
since n1 is incremented in a loop and not related to n1_add, I decided to just split the associated data in half.

gilles-peskine-arm replied:
Splitting in half is not a good test because the test data lengths don't vary much. The reason `n1` is incremented in a loop is that it gives good test coverage: at the very least, we want the split to happen at 16*a+b for all b=0..15, because there's a risk of bugs related to how far the split is into a 16-byte block.

Correlating `n1` with `n1_add` is not ideal but better than `add->len/2`. Actually I think `add->len - n1` would give slightly better coverage (keeping it within the range `0`..`add->len` of course).

mstarzyk-mobica commented on include/mbedtls/gcm.h:
```diff
@@ -69,6 +69,7 @@ typedef struct mbedtls_gcm_context
     int mode;                             /*!< The operation to perform:
                                                #MBEDTLS_GCM_ENCRYPT or
                                                #MBEDTLS_GCM_DECRYPT. */
+    uint8_t ad_data_finished;
```
a flag ment to be checked during first call to `mbedtls_gcm_update` or mbedtls_gcm_finish`.

gilles-peskine-arm replied:
How about using `ctx->len==0`? This tells you whether `update` has been called with a nonzero length, which is slightly different (in case `update` is first called with length=0) but gives the same result in the end. By leveraging this existing information, you don't need the code to update the flag, just the code to finish authenticating the AD when `ctx->len == 0 && ctx->add_len % 16 1= 0` at the beginning of `update` or `finish`.

mstarzyk-mobica replied:
ah! great! exactly what I needed :)

mstarzyk-mobica commented on library/gcm.c:
```diff
@@ -337,23 +337,36 @@ int mbedtls_gcm_update_ad( mbedtls_gcm_context *ctx,
+
+    if (offset)
+    {
+        use_len = 16 - offset;
```
magic value 16 is used commonly here, probably a good idea to just use the length of `ctx->buf`

gilles-peskine-arm replied:
The length of the buffer is not a good value here: it's the maximum supported block length, not the block length for the algorithm. If we ever support algorithms with a different length (which is not going to happen anytime soon), having `sizeof(buf)` would be worse than `16` since 16 is clearly a magic number but `sizeof(buf)` is correct when initializing and wiping buffers. So please keep 16 as a magic value as in the existing code.

gilles-peskine-arm commented on library/gcm.c:
```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
         return( MBEDTLS_ERR_GCM_BAD_INPUT );
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
```
Style:
```suggestion
    if( ctx->len == 0 && ctx->add_len % 16 != 0 )
```
(No space before `(` per style guide. No implicit comparison with 0 per style guide. Not putting redundant parentheses here is my personal preference, I think they mislead the reader into thinking that something unusual is going on, but feel free to leave them if you prefer.)

gilles-peskine-arm commented on library/gcm.c:
```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
+    {
```
Partly on me when I worked on `update` but it's getting worse: please add comment in `gcm.c` (in the documentation of `mbedtls_gcm_context` would be better but it would conflict with #4511 so let's avoid that so let's put it at the top of the implementation of `mbedtls_gcm_update_ad` maybe?) explaining what `buf` contains at various stages:
* When `add_len % 16 == 0` and `len == 0`: initial state.
* When `add_len % 16 != 01 and `len == 0`: the first `add_len % 16` bytes of `buf` have a partial AD block xored in and not yet multiplied in.
* When `len != 0`: the first `add_len % 16` bytes of `buf` have partial ciphertext xored in and not yet multiplied in.

gilles-peskine-arm reviewed (changes requested):
The code looks correct to me. A few documentation changes are needed, and there's an edge case that needs better testing.

mstarzyk-mobica commented on library/gcm.c:
```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
+    {
```
- [ ]  add a test function with non-empty AD and empty encrypted data, checking both encryption and decryption, and calling mbedtls_gcm_update(0) 0 or 1 time based on the value of a parameter.

mstarzyk-mobica replied:
Hm. I'm not sure I did what you expected. I added a test in the test data, to cover this scenario. Is it ok?

gilles-peskine-arm replied:
I meant to test the scenario where the functions are called in the sequence above. There is currently no test scenario where `mbedtls_gcm_update_ad` is called at least once, but `mbedtls_gcm_update` is not called at all. When we test, even if the plaintext/ciphertext is empty, we call `mbedtls_gcm_update` with a length of 0. It might well make a difference whether `mbedtls_gcm_update` is called at all or not: a plausible bug would be that the last partial block of AD is processed correctly only if `mbedtls_gcm_update` is called at least once, or only if `mbedtls_gcm_update` is called at most once, or only if `mbedtls_gcm_update` is not called at all.

So we need a new test function, which is specialized to empty plaintext/ciphertext, where the AD is a test parameter and where the number of times to call `mbedtls_gcm_update` is a parameter or a loop index. This must be called for at least AD lengths 0, 16 and some number not multiple of 16; for both encryption and decryption; with `mbedtls_gcm_update` called 0, 1 and 2 times.

gilles-peskine-arm replied:
And similarly we should have tests with empty AD where `mbedtls_gcm_update_ad` is not called at all. Given the current structure of the code, a bug in this case is not likely, but we should test it in case the structure of the code changes.

ronald-cron-arm commented on tests/suites/test_suite_gcm.function:
```diff
@@ -16,15 +16,19 @@ static int check_multipart( mbedtls_gcm_context *ctx,
     int ok = 0;
     uint8_t *output = NULL;
     size_t n2 = input->len - n1;
+    size_t n1_add = n1 < add->len ? add->len - n1 : add->len;
```
I am not a big fan of this correlation of the split of additional data and data to encrypt/decrypt. Please add another argument for the split of additionnal data and in the caller functions change to two nested for loops.

gilles-peskine-arm commented on library/gcm.c:
```diff
@@ -332,28 +332,53 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
 
-
+/**
+ * mbedtls_gcm_context::buf contains different data type, depending
```
It's not “different data type”: it's always the partial state of the computation of the authentication tag. The breakdown is what partially processed input is at different stages.

gilles-peskine-arm commented on library/gcm.c:
```diff
@@ -332,28 +332,53 @@ int mbedtls_gcm_starts( mbedtls_gcm_context *ctx,
+ * mbedtls_gcm_context::buf contains different data type, depending
+ * on the values of mbedtls_gcm_context::::add_len and
+ * mbedtls_gcm_context::len:
+ *     * When add_len % 16 == 0 and len == 0: initial state.
```
Looking at this again I think it would be best explained with four cases:
* If `len == 0 && add_len % 16 == 0`: the first `add_len % 16 == 0` bytes have a partial block of AD that has been xored in but not yet multiplied in.
* If `len == 0 && add_len % 16 == 0`: the authentication tag is correct if the data ends now.
* If `len % 16 != 0`: the first `len % 16 == 0` bytes have a partial block of ciphertext that has been xored in but not yet multiplied in.
* If `len > 0 && len % 16 == 0`: the authentication tag is correct if the data ends now.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.aes128_de.data:
```diff
@@ -670,6 +670,10 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_decrypt_and_verify:MBEDTLS_CIPHER_ID_AES:"659b9e729d12f68b73fdc2f7260ab114":"fd0732a38224c3f16f58de3a7f333da2ecdb6eec92b469544a891966dd4f8fb64a711a793f1ef6a90e49765eacaccdd8cc438c2b57c51902d27a82ee4f24925a864a9513a74e734ddbf77204a99a3c0060fcfbaccae48fe509bc95c3d6e1b1592889c489801265715e6e4355a45357ce467c1caa2f1c3071bd3a9168a7d223e3":"459df18e2dfbd66d6ad04978432a6d97":"ee0b0b52a729c45b899cc924 ...
 
+AES-GCM NIST Validation (AES-128,128,0,256,128) #0 [#0]
```
There are already plenty of test cases with empty ciphertext and non-empty AD, so I don't think the two new test cases are useful.

----

In case they do turn out to be useful:

Please don't add `[#N]` annotations on test cases and avoid adding `#N`. `#N` is a bad historical habit of using numbers rather than useful descriptions. `[#N]` annotations were added automatically when test cases had identical descriptions which made the test reporting hard to use. In both cases, we eventually want to replace those by meaningful descriptions (except in rare cases where we're just testing crypto with some randomish data and there's nothing informative about the difference between randomish data 1 and randomish data 2). If the goal of this test case is to test with empty ciphertext and non-empty AD, the test description should say so. 

Did this test data actually come from a NIST validation suite? I can't find `ff7ad50df57d635839acf143af43f002` in 
http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip .

mstarzyk-mobica replied:
I was not aware we should pick test vectors from some special place. I guess I'll remove them anyways, since (as you pointed it out) they don't cover increase the coverage.

gilles-peskine-arm replied:
We don't have to pick test vectors from a special place, but 1. we do need to know why we believe that they are correct and 2. if the description says “NIST Validation” then they should indeed come from a NIST validation suite.

mstarzyk-mobica commented:
1. rebased with removal of redundant commit
```
commit 8fbee92d3cc141d7ca04c7b93e4a854e1d135b1f
Author: Mateusz Starzyk <mateusz.starzyk@mobica.com>
Date:   Wed Jun 9 16:26:55 2021 +0200

    Add gcm tests for empty ciphertext and non-empty AAD.
```

2. rebased to `development` fixing the merge conflict

mstarzyk-mobica commented on tests/suites/test_suite_gcm.aes128_de.data:
```diff
@@ -670,6 +670,30 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 
+AES-GCM NIST Empty ciphertext. Non-empty AD. Ciphertext updates: 0
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_cipher:MBEDTLS_CIPHER_ID_AES:"63c3f81500746eaf383fe3975d84f849":"0799d4152fd73c1604b4610cf7171fe1":"cb8248e5f904cc9ccccf6f273fe621eee1b4d7ed98480f9e806a48b84e2d6a733772ecf8fb7fe91805715cddab2b462b89f6e6c7cf873f65031f13c357d5f57b00b7c391c39e78ad1ed94be236ca0ae316bce11bc33c5d701fdfc58abbe918b9c42f7b3d6e89d46f9784b388a6e6daf47730b9fa665d755a17e89932fa669c44":"c53d01e53ee ...
```
is it ok to have same test vector for all of these test cases?

gilles-peskine-arm replied:
Yes. The goal is to test the state machine of the GCM calculation. The exact data doesn't matter, only the lengths. For the same reason, it's ok to test only with AES128 and not the other block ciphers.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.function:
```diff
@@ -123,11 +124,16 @@ void gcm_encrypt_and_tag( int cipher_id, data_t * key_str,
+
+            for( n1_add = 0; n1_add <= add_str->len; n1_add += 1 )
+            {
+                mbedtls_test_set_step( n1_add );
```
The point of the test step is to convey the conditions in which a test failure happens. When tests have no loop or auxiliary functions, the test case name and the line number are enough. The step number provides additional information when the same code line runs more than once for a test case. With a simple loop, the iteration number is enough information. With a nested loop, the step number should encode the step number from both loops. Here, with a test failure, we'd only see the problematic value of `n1_add` and not the problematic value of `n1`.

The step number is printed out in decimal. Following a precedent in `tests/suites/test_suite_ssl.function`:
```
mbedtls_test_set_step( n1 * 10000 + n1_add );
```

Same in the decrypt function.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.function:
```diff
@@ -123,11 +124,16 @@ void gcm_encrypt_and_tag( int cipher_id, data_t * key_str,
-                                  n1 ) )
-                goto exit;
+
+            for( n1_add = 0; n1_add <= add_str->len; n1_add += 1 )
```
The nested loops make the GCM tests a bit slower than I'd like, and we could cut out a lot of values of (n1, n1add) without reducing coverage. But let's optimize later. Filed as https://github.com/ARMmbed/mbedtls/issues/4670.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.function:
```diff
@@ -196,6 +267,110 @@ exit:
+
+    mbedtls_gcm_init( &ctx );
+
+    TEST_ASSERT( mbedtls_gcm_setkey( &ctx, cipher_id, key_str->x, key_str->len * 8 ) == init_result );
```
This function doesn't to exercise the case where setkey fails, so `init_result` is not useful. Also goes for the other 3 functions.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.aes128_en.data:
```diff
@@ -670,6 +670,30 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_encrypt_and_tag:MBEDTLS_CIPHER_ID_AES:"fe481476fce76efcfc78ed144b0756f1":"246e1f2babab8da98b17cc928bd49504d7d87ea2cc174f9ffb7dbafe5969ff824a0bcb52f35441d22f3edcd10fab0ec04c0bde5abd3624ca25cbb4541b5d62a3deb52c00b75d68aaf0504d51f95b8dcbebdd8433f4966c584ac7f8c19407ca927a79fa4ead2688c4a7baafb4c31ef83c05e8848ec2b4f657aab84c109c91c277":"1a2c18c6bf13b3b2785610c71ccd98ca":"b0ab3cb5256575774b8242b89ba ...
 
+AES-GCM NIST Empty ciphertext. Non-empty AD. Ciphertext updates: 0
```
We know that 16 is a critical value. So please have both a set of test cases where the AD length is a nonzero multiple of 16, and a set where it isn't a multiple of 16. Same with the empty AD tests and the plain/cipher-text length.

gilles-peskine-arm commented on tests/suites/test_suite_gcm.function:
```diff
@@ -196,6 +267,110 @@ exit:
+    mbedtls_gcm_free( &ctx );
+}
+/* END_CASE */
+
```
One more edge case that needs to be tested is when both AD and plaintext/ciphertext are empty, i.e. the sequence is setkey, finish.

gilles-peskine-arm reviewed (changes requested):
A few remaining issues in the tests. Apart from that looks good to me.

ronald-cron-arm commented on tests/suites/test_suite_gcm.aes128_de.data:
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_cipher:MBEDTLS_CIPHER_ID_AES:"63c3f81500746eaf383fe3975d84f849":"0799d4152fd73c1604b4610cf7171fe1":"cb8248e5f904cc9ccccf6f273fe621eee1b4d7ed98480f9e806a48b84e2d6a733772ecf8fb7fe91805715cddab2b462b89f6e6c7cf873f65031f13c357d5f57b00b7c391c39e78ad1ed94be236ca0ae316bce11bc33c5d701fdfc58abbe918b9c42f7b3d6e89d46f9784b388a6e6daf47730b9fa665d755a17e89932fa669c44":"c53d01e53ee ...
+
+AES-GCM NIST - empty ciphertext, AD length: 90 bytes, ciphertext updates: 0
```
Is the reference to NIST correct here ? For the three tests before I can see that there is a corresponding NIST test case before by not for this one. Related question, where does this test data come from? Ditto for the "en" data.

mstarzyk-mobica replied:
Test data comes from http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip

ronald-cron-arm replied:
Thanks for the reference. Wouldn't it better to say "NIST CAVS 14.0" to identify the source ?

gilles-peskine-arm replied:
It would be nice to identify the source precisely, with the URL, in the commit message. Completely precise identification is a bit too much information in the commit description, which has to fit in 66 characters to be displayed fully.

ronald-cron-arm replied:
Just asking as I don't really know how it works with NIST and its test vectors. Isn't the case that the URL used here could in the future point to a new version of the test vectors which does not include the test vectors added here anymore?

gilles-peskine-arm replied:
Yes. For an unambiguous reference, you need both the URL and some version information. Some indication of the download date, such as an implicit date of “a few days before the commit”, is good enough. A version number is definitely better when there is one.

mstarzyk-mobica replied:
Should I edit the commit messages? (rebase + force push required)

ronald-cron-arm replied:
Editing the commit message is not an issue from our reviewer perspective as we can check that the diff of the force-push is empty. Thus please improve the commit message. Otherwise the NIST files you pointed out start with the header:

 CAVS 14.0
# GCM Decrypt with keysize 128 test information
# Generated on Fri Aug 31 11:28:04 2012

CAVS 14.0 seems to be the version and I think it would be a good idea to mention it in the test description, thus:
`AES-GCM NIST CAVS 14.0 - ... `

ronald-cron-arm commented on tests/suites/test_suite_gcm.aes128_de.data:
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_empty_ad:MBEDTLS_CIPHER_ID_AES:"c58583f6479d9bc9f1bffddefee66e59":"cee448b48d3506ff3ecc227a87987846":"564a9f700cbc1f895e4f4fa6426f73b4956896a15e6127e7560d74e3fd0b980d2ee45b7a6a3884fa613d91d13921e3f90967d7132bdafcd146dd8ff7147ed1964c2bdb3e12f4133d3dbbc3bf030ff37b1d2147c493ce885068d9ba5bebae24903aaac004aa0ab73fe789e4150e75ddc2bde2700db02e6398d53e88ac652964ac":"361fc2896d7ee98 ...
+
+AES-GCM NIST - empty AD, ciphertext length: 51 bytes, AD updates: 0
```
ditto, NIST? Where does the data come from? Ditto for "en" data.

ronald-cron-arm commented on tests/suites/test_suite_gcm.aes128_de.data:
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
+
+AES-GCM NIST - empty AD, empty ciphertext
+depends_on:MBEDTLS_AES_C
+gcm_decrypt_and_verify_no_ad_no_cipher:MBEDTLS_CIPHER_ID_AES:"cf063a34d4a9a76c2c86787d3f96db71":"113b9785971864c83b01c787":"72ac8493e3a5228b5d130a69d2510e42"
```
Ditto NIST? Origin? "en" data.

ronald-cron-arm reviewed (changes requested):
A few remaining things.

mstarzyk-mobica commented on tests/suites/test_suite_gcm.function:
```diff
@@ -57,6 +61,94 @@ exit:
+    TEST_EQUAL( 0, mbedtls_gcm_starts( ctx, mode,
+                                       iv->x, iv->len ) );
+
+    for( n = 0; n < ad_update_count; n += 1 )
```
`+= 1 ` was actually already used in this file ;)

gilles-peskine-arm commented on tests/suites/test_suite_gcm.aes128_en.data:
```diff
@@ -670,6 +670,58 @@ AES-GCM NIST Validation (AES-128,128,1024,1024,32) #2 [#2]
 depends_on:MBEDTLS_AES_C
 gcm_encrypt_and_tag:MBEDTLS_CIPHER_ID_AES:"fe481476fce76efcfc78ed144b0756f1":"246e1f2babab8da98b17cc928bd49504d7d87ea2cc174f9ffb7dbafe5969ff824a0bcb52f35441d22f3edcd10fab0ec04c0bde5abd3624ca25cbb4541b5d62a3deb52c00b75d68aaf0504d51f95b8dcbebdd8433f4966c584ac7f8c19407ca927a79fa4ead2688c4a7baafb4c31ef83c05e8848ec2b4f657aab84c109c91c277":"1a2c18c6bf13b3b2785610c71ccd98ca":"b0ab3cb5256575774b8242b89ba ...
 
+AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ciphertext updates: 0
```
For future reference, our test framework scripts truncate test case descriptions at 66 characters, so it's better if they fit in 66 characters, and annoying when the first 66 characters aren't unique.
```
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 128 bytes, ci  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty plaintext, AD length: 90 bytes, cip  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 128 bytes, AD  PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
AES-GCM NIST CAVS 14.0 - empty AD, plaintext length: 51 bytes, AD   PASS
```
It's a preexisting problem in many test suites, and we're pressed for time, so I'm not going to block the PR on this.
